### PR TITLE
Fix Cadence's unsafe random

### PR DESCRIPTION
**Pull request by tarakby**
Closes https://github.com/onflow/cadence/issues/2691

Emulator now supports Cadence random function(s). The randoms are currently deterministic and equal for all blocks.

______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to GitHub issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-emulator/blob/master/CONTRIBUTING.md#styleguides)
- [x] Re-reviewed `Files changed` in the GitHub PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: emulator/blockchain.go [hunk 1]
```diff
@@ -530,6 +530,19 @@ func (h CadenceHook) Run(_ *zerolog.Event, level zerolog.Level, msg string) {
 	}
 }
 
+// `dummyEntropyProvider“ implements `environment.EntropyProvider`
+// which provides a source of entropy to fvm context (required for Cadence's randomness).
+type dummyEntropyProvider struct{}
+
+var dummySource = make([]byte, 32)
```
Comment (tarakby) [inline emulator/blockchain.go:537]: This causes randomness on emulator to be:
 1. deterministic
 2. same for all txs/blocks
 
 Feel free to update this to be any behaviour you prefer.

**Comment by janezpodhostnik** [comment]
> does this break mainnet/testnet transaction replay ( simulation ) ? Is there a way we can solve this ? maybe by getting block (transaction ) random seed from archive nodes or similar?

yes, it does break it... but, like you said, we can fix it by exposing the random beacon. 
@tarakby is there anyway to access the random beacon for sealed blocks?

**Comment by bluesign** [comment]
> > does this break mainnet/testnet transaction replay ( simulation ) ? Is there a way we can solve this ? maybe by getting block (transaction ) random seed from archive nodes or similar?
> 
> yes, it does break it... but, like you said, we can fix it by exposing the random beacon. @tarakby is there anyway to access the random beacon for sealed blocks?

maybe we can add to transaction result as random seed.

**Comment by tarakby** [comment]
> @tarakby is there anyway to access the random beacon for sealed blocks?

The issue with the source of randomness is that it is part of the block payload data not the execution state. Archive nodes and other nodes in general are not required to store these data for all the history (most of them only store data since the root  block of the current spork only, but not before that).

> maybe we can add to transaction result as random seed.

Yes this is a good idea - We can restrict it only to transactions that used Cadence's random function, since most of the transactions do not need it. 

I have no idea how to plug that seed into the emulator later when replaying a transaction. Does this PR need further changes to accommodate that?

**Review by bluesign — APPROVED** [review]

**Comment by bluesign** [comment]
I think we can merge this in the mean time, for sure random seed in tx result would take some time

**Comment by btspoony** [comment]
Hi @turbolent, Approximately when can this PR be included in a certain release of `flow-cli`?

**Comment by turbolent** [comment]
@btspoony I only assist with reviews in this project. Please open an issue in this and the CLI repo. Thank you!

**Pull request merged**